### DISCUSSION TITLE
ECS Fargate provider confguration errors during terraform init

### DISCUSSION THREAD

**Discussion by rhoboat** [gruntwork-io/knowledge-base#192] [Help] [answered]
I'm using ECS Fargate modules. When running **`terraform init`**, I get this error.

```
Downloading git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git?ref=v0.73.2 for fargate_cluster...
- fargate_cluster in .terraform/modules/fargate_cluster/modules/services/ecs-fargate-cluster
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that Terraform can determine which modules and
providers need to be installed.
╷
│ Error: Module module.fargate_service.module.route53_health_check contains provider configuration
│ 
│ Providers cannot be configured within modules using count, for_each or depends_on.
```
╵Terraform v1.0.11

I'm using:

```
module "fargate_cluster" {
  source = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/services/ecs-fargate-cluster?ref=v0.73.2"

...
}

module "fargate_service" {
  source = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/services/ecs-service?ref=v0.73.2"

...
}
```

In case this might be related to the Terraform AWS provider v4.0.0, my `.terraform.lock.hcl` file
```
provider "[registry.terraform.io/hashicorp/aws](http://registry.terraform.io/hashicorp/aws)" {
  version     = "3.74.1"
  constraints = ">= 2.6.0, >= 2.68.0, ~> 3.23, >= 3.27.0"
  hashes = [
```

How do I get past this error?

**Comment by rhoboat** [comment]
My gut feeling is to make sure there isn't some hidden provider configuration that's overriding the lock file. What's the file path for the `.terraform.lock.hcl` that you posted? It might also be something within our modules trying to configure a provider internally (checking this)... Maybe it's something to do with [this](https://github.com/gruntwork-io/terraform-aws-monitoring/blob/master/modules/alarms/route53-health-check-alarms/main.tf#L14-L26)?

**Reply by KOConchobhair** [reply]
its a typical folder structure (no terragrunt involved). the modules are defined in main.tf obviously.
```
.
├── .terraform
├── .terraform.lock.hcl
├── backend.tf
├── main.tf
├── outputs.tf
├── variables.tf
└── versions.tf
```

versions.tf had
```
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.23"
    }
  }
```
changed it to exclude 4.0
```
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.23, < 4.0"
    }
  }
```
ran `terraform init -upgrade` still got the error. so its not related to v4 aws provider
```
╷
│ Error: Module module.fargate_service.module.route53_health_check contains provider configuration
│ 
│ Providers cannot be configured within modules using count, for_each or depends_on.
╵
```

**Comment by KOConchobhair** [comment]
looks related to this issue https://github.com/gruntwork-io/terraform-aws-monitoring/issues/183
which was fixed in this pr https://github.com/gruntwork-io/terraform-aws-monitoring/pull/184
released in terraform-aws-monitoring https://github.com/gruntwork-io/terraform-aws-monitoring/releases/tag/v0.26.1
aws service catalog (master) is using `terraform-aws-monitoring.git//modules/alarms/route53-health-check-alarms?ref=v0.30.5` so the fix is available...
but the aws service catalog's ecs-service does not expose the necessary variables from `route53-health-check-alarms` module https://github.com/gruntwork-io/terraform-aws-monitoring/blob/830c40f2863d690ebe0b6ba9a22b989e805d9bbe/modules/alarms/route53-health-check-alarms/vars.tf#L46

so you'll need to decide which of these variables (all of them?) you want to expose in all the services in the service catalog that rely on `route53-health-check-alarms`...ill file an [issue](https://github.com/gruntwork-io/terraform-aws-service-catalog/issues/1201). This is blocking me pretty good right now!

**Comment by KOConchobhair** [comment]
hmm, going back over the error message `Providers cannot be configured within modules using count, for_each or depends_on.`

it turns out that this is the root cause...removing depends_on makes the error go away. but now I cannot enforce the order. hopefully its not necessary (i found it from a gruntwork example though)

```hcl
module "fargate_cluster" {
  source = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/services/ecs-fargate-cluster?ref=v0.73.2"
  
  cluster_name = "test"
}

module "fargate_service" {
  source = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/services/ecs-service?ref=v0.73.2"
  
  service_name     = "test"
  ecs_cluster_arn  = module.fargate_cluster.arn
  ecs_cluster_name = module.fargate_cluster.name
  # Make sure all the ECS cluster resources are deployed before deploying any ECS service resources. This is also
  # necessary to avoid issues on 'destroy'.
  depends_on = [module.fargate_cluster]
}
```

**Reply by Etiene** [reply]
Hello @KOConchobhair,  using the `module.fargate_cluster.arn` and `module.fargate_cluster.name` variables should already enforce an order through implicit dependencies, unless you are having issues with terraform's implicit graph generation it shouldn't be necessary. More info here: https://technotrampoline.com/articles/how-does-terraform-depends-on-work/ and here: https://blog.jcorioland.io/archives/2020/05/12/terraform-implicit-explicit-dependencies-between-resources.html

Is this still blocking you?

**Reply by KOConchobhair** [reply]
no, I removed `depends_on` since it wasn't necessary in the first place due to the implicit dependencies. HOWEVER...

A) I got that from a gruntwork example: https://github.com/gruntwork-io/terraform-aws-ecs/blob/307cc8f9f973cc6aa52c6c085358af205aca3e8b/examples/docker-fargate-service-with-efs-volume/main.tf#L72 
so that is where the original confusion came from in the first place. so you may want to fix that example code.

B) all of this indirectly uncovered a bug due to nested providers, as mentioned here: https://github.com/gruntwork-io/terraform-aws-monitoring/issues/244
nested providers need to be removed from modules otherwise it restricts how they can be used (can't use `depends_on` when the time comes where it will be actually necessary)

**Comment by rhoboat** [comment, marked as answer]
@KOConchobhair The issue is that the provider is being specified in the monitoring module, so the modules that use that cannot specify `depends_on`. We need to remove nested providers. We shouldn't [specify the provider here](https://github.com/gruntwork-io/terraform-aws-monitoring/blob/3c8d171/modules/alarms/route53-health-check-alarms/main.tf#L18-L30). I've [filed an issue](https://github.com/gruntwork-io/terraform-aws-monitoring/issues/244) for this.

I think the workaround of removing `depends_on` is reasonable. The `fargate_service` depends on the outputs from `fargate_cluster` within the configuration, so Terraform will figure out that dependency and apply the cluster before the service.
